tink/internal/controllers/template/controller_internal_test.go

```diff
@@ -0,0 +1,255 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package controllers contains controllers for Tinkerbell.
+package template
+
+import (
+	"context"
+	"testing"
+
+	. "github.com/onsi/gomega"
+	tinkv1alpha1 "github.com/tinkerbell/cluster-api-provider-tinkerbell/tink/api/v1alpha1"
+	tinkfake "github.com/tinkerbell/cluster-api-provider-tinkerbell/tink/internal/client/fake"
+	"github.com/tinkerbell/tink/protos/template"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/utils/pointer"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const helloWorldTemplate = `version: "0.1"
+name: hello_world_workflow
+global_timeout: 600
+tasks:
+  - name: "hello world"
+    worker: "{{.device_1}}"
+    actions:
+      - name: "hello_world"
+        image: hello-world
+        timeout: 60`
+
+func TestTemplateReconciler_reconcileDelete(t *testing.T) {
+	g := NewWithT(t)
+	scheme := runtime.NewScheme()
+	now := metav1.Now()
+
+	g.Expect(tinkv1alpha1.AddToScheme(scheme)).To(Succeed())
+
+	tests := []struct {
+		name     string
+		in       *tinkv1alpha1.Template
+		tinkObjs []*template.WorkflowTemplate
+		want     ctrl.Result
+		wantErr  bool
+	}{
+		{
+			name: "successful delete by id",
+			in: &tinkv1alpha1.Template{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:              "test",
+					DeletionTimestamp: &now,
+					Finalizers:        []string{tinkv1alpha1.TemplateFinalizer},
+					Annotations: map[string]string{
+						tinkv1alpha1.TemplateIDAnnotation: "testId",
+					},
+				},
+			},
+			tinkObjs: []*template.WorkflowTemplate{
+				{
+					Id:   "testId",
+					Name: "test",
+				},
+			},
+			want:    ctrl.Result{},
+			wantErr: false,
+		},
+		{
+			name: "template doesn't exist in tinkerbell",
+			in: &tinkv1alpha1.Template{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:              "test",
+					DeletionTimestamp: &now,
+					Finalizers:        []string{tinkv1alpha1.TemplateFinalizer},
+					Annotations: map[string]string{
+						tinkv1alpha1.TemplateIDAnnotation: "testId",
+					},
+				},
+			},
+			tinkObjs: []*template.WorkflowTemplate{},
+			want:     ctrl.Result{},
+			wantErr:  false,
+		},
+	}
+	for i := range tests {
+		tt := tests[i]
+		t.Run(tt.name, func(t *testing.T) {
+			g := NewWithT(t)
+
+			fakeTemplateClient := tinkfake.NewFakeTemplateClient(tt.tinkObjs...)
+
+			r := &Reconciler{
+				Client:         fake.NewFakeClientWithScheme(scheme, tt.in.DeepCopy()),
+				TemplateClient: fakeTemplateClient,
+				Log:            log.Log,
+				Scheme:         scheme,
+			}
+
+			got, err := r.reconcileDelete(context.Background(), tt.in)
```

Comment:
I guess we could just call `Reconcile` here and avoid having this test as internal? As we set deletion timestamp anyway.